src/counters.ts

```diff
@@ -0,0 +1,711 @@
+	}
+
+	private async writeCollectionData(): Promise<void> {
+		// data, simple moving average data and exponential moving average data are all written to different files
+		// so we kick them all off and then wait for all of them to complete.
+		let promises: Promise<void>[] = [];
+		let file = `${this.getOutputFileName()}_data.json`;
+		promises.push(writeFileWithPromise(file, jsonDump(this.collection)));
+		if (this.includeMovingAverages) {
+			file = `${this.getOutputFileName()}_sma_data.json`;
+			promises.push(writeFileWithPromise(file, jsonDump(this.smaOver4Collection)));
+			file = `${this.getOutputFileName()}_ema_data.json`;
+			promises.push(writeFileWithPromise(file, jsonDump(this.emaOver4Collection)));
+		}
+		await Promise.all(promises);
+	}
+
+
+	private getOutputFileNameForProcess(proc: ProcessInfo = undefined): string {
+		let file: string = this.getOutputFileName();
+		if (proc) {
+			file = proc.pid > 0 ? `${file}__${proc.name}_${proc.pid}` : `${file}__${proc.name}`;
+		}
+		return file.replace('.', '_');
+	}
+
+	private getOutputFileName(): string {
+		return path.join(this.outputDirectory, this.name);
+	}
+
+	/**
+	 * This method resets this object. It stops any ongoing collection as well as clears any collected data.
+	 */
+	public async reset(): Promise<void> {
+		await this.stop();
+		this.collection.clear();
+		this.smaOver4Collection.clear();
+		this.emaOver4Collection.clear();
+	}
+
+	//processes to track for counters  
+	private processesToTrack: ProcessInfo[] = [];
+
+	// timer for updating processesToTrack in case the process at the root of {@link pid} has changed.
+	private processesToTrackTimer: NodeJS.Timer = null;
+
+	// timer for collecting the statistics of the  processes that we are tracking.
+	private countersTimer: NodeJS.Timer = null;
+
+	// Promise for updating processesToTrack in case the process at the root of {@link pid} has changed.
+	private processesToTrackUpdatePromise: Promise<void> = null;
+
+	// Promise for collecting the statistics of the  processes that we are tracking.
+	private countersCollectionPromise: Promise<void> = null;
+
+	// flag for updating processesToTrack in case the process at the root of {@link pid} has changed.
+	private processesToTrackUpdateInProgress: boolean = false;
+
+	// flag for collecting the statistics of the  processes that we are tracking.
+	private countersCollectionInProgress: boolean = false;
+
+	private async stopPopulatingProcessInfos(): Promise<void> {
+		clearInterval(this.processesToTrackTimer);
+		if (this.processesToTrackTimer) {
+			this.processesToTrackTimer.unref();
+			this.processesToTrackTimer = null;
+		}
+		await this.processesToTrackUpdatePromise;
+
+	}
+
+	private async stopCollecting(): Promise<void> {
+		await this.countersCollectionPromise;
+		clearInterval(this.countersTimer);
+		if (this.countersTimer) {
+			this.countersTimer.unref();
+			this.countersTimer = null;
+		}
+		await this.countersCollectionPromise;
+	}

```

Comment:
Why do you await on this twice? Just doing it once after stopping the timer should be enough.

Reply:
yeah - no need, it seems like some copy and paste bug. Will remove.